### ISSUE TITLE
cleanup cli parsing / subcommands

### ISSUE THREAD

**Issue by Dieterbe** [naggie/dstask#96]
Here's my issue with the current state of things:

* cmd/dstask/main.go is unwieldy. we shouldn't have to hardcode all the cases to check every command string and link it to the right function.
* help.go is unwieldy, incomplete (see #83), and the doc strings are detached from the code they correspond to. I think this is an antipattern in go. whereas in some other languages you it may be recommended to have a separate file for all helptext etc, in go it seems more common to bundle things together if they belong together, even if they are of different "types" (e.g. help text and code). it would also make it obvious which commands lack a help message.  also the help text is manually maintained, it would be nice if we could auto-generate at least portions of it (this is common with flag args, though we don't use those. I actually like the cmdline args only approach). Also help.go shouldn't live in the root of the repo, as it pertains only to the dstask cli program.
* `dstask --help` doesn't work (though `dstask help` does but that's counter-intuitive. the flag form should work and `dstask help` looks more like a filter)
* the ParseCmdLine routine is weird. While different commands have different expected commandlines, the same function is called for all of them and it somehow works for all these different scenarios? this code is hard to follow.  it seems like it would make more sense to first look at the invoked command and run it.  each command would do (or at least declare) its own commandline parsing (possibly lots of code reuse amongst them, but the point is that it would be more explicit what each one needs, and it would be obvious to see which commands use the same expected commandline). While we're at it, i would propose a more step-wise parsing (e.g. we expect an id, look for it, then we expect  a filter, parse that, then notes, parse that, etc. rather than running the same loop over and over and setting all kinds of state flags)
* completions.. can probably be auto-generated with certain libraries. though i'm not sure.
* pluggable subcommand support ( @dontlaugh [has mentioned he likes how terraform does it](https://github.com/naggie/dstask/pull/81#issuecomment-731849027))

here's a few commonly used libraries:

* https://github.com/spf13/cobra : 13.5k lines - this one has a lot of features and is perhaps a bit too much
* https://github.com/urfave/cli. 14k lines - i thought this would be way more lightweight and am actually surprised by the LOC
* https://github.com/mitchellh/cli: 3200 loc. (add another 1k for the dependency on the go-radix library). note that this is the lib used by terraform and other hashicorp projects.

anyway, i don't have a thorough understanding of our needs yet. we could probably clean up what we have with DIY code and that should leave us in a pretty good spot.  @dontlaugh's preferred plugin system could also be done using DIY code. we probably don't really need 3rd party libs for quite a while.

**Comment by dontlaugh** [comment]
Thank you for bringing these maintainability issues up. I believe that the maintainability situation can be improved. I also think we will need a "custom" parser.

Consider this command line

```
dstask project:move +bills Cancel internet service / website is http://myhorribleisp.net
``` 

In one go, we add a task with 4 types of attributes: text, tag, project, note. My guess is that urfave/cli and cobra would have trouble accommodating these. The standard cli frameworks have types like: positional arg, flags that take values, flags that don't take values. We could make it work, but I think we'd (ironically) end up writing parsing code after the inital cli library has done it's own sort of parsing.

The dstask command line is a DSL. It's so simple that it's basically a data structure, but it still needs parsing. 

In the codebase, we do tokenizing/some parsing in `ParseCmdLine` and then a little more parsing in commands.go (the giant switch statement still has elements that might be considered parsing).

So perhaps a way to improve maintainability is to really lean into that concept, and bring to bear the tools for dealing with DSLs. 

We could:
* keep refining our existing parsing code, but adopt naming conventions and code structure that is more in line with a traditional tokenize -> parse -> execute interpreter flow. We could work towards a code design that has some of the niceties of cobra, such as co-locating help text with where commands are declared
* Use a parser generator (perhaps [this one](https://github.com/pointlander/peg) or [this other one](https://github.com/mna/pigeon)). This might seem like overkill for such a small language, but it would make the language easier to evolve.

**Comment by Dieterbe** [comment]
> dstask project:move +bills Cancel internet service / website is http://myhorribleisp.net

this gets interpreted as a search query. this adds the task:

> dstask add project:move +bills Cancel internet service / website is http://myhorribleisp.net

but i agree with your overall point, it seems more sensible to spend time doing our own parsing (possibly with a helper library) rather than evaluating and prototyping with libraries such as cli or cobra.
in one of my other side projects I've used [toki](https://github.com/taylorchu/toki). it's quite simplistic IIRC but it might be just the right amount of simplistic. I'm not very well versed in the terminology here, but I still ended up writing my own "expect this token, now check if we have this or that token" etc logic though (this is referred to as parsing, i suppose?). i don't think it allows you to express a language (as in, multiple branches of syntax that can optionally be taken, possibly nested, optional sections, etc. I guess that would be a parser generator?). Assuming i got the terminology right, then i think doing our own parsing can go a long way before we need a parser generator. does a generator have other advantages, such as the ability to generate help text from the language specification? If we DIY, we'd have to keep the help text in sync manually. Not a huge deal though.

**Comment by naggie** [comment]
> The ParseCmdLine routine is weird. While different commands have different expected commandlines, the same function is called for all of them and it somehow works for all these different scenarios? 

An yes, it does "somehow" work for all scenarios so far; there could be additional logic to check for invalid invocations though I guess.

> this code is hard to follow. it seems like it would make more sense to first look at the invoked command and run it. each command would do (or at least declare) its own commandline parsing (possibly lots of code reuse amongst them, but the point is that it would be more explicit what each one needs, and it would be obvious to see which commands use the same expected commandline). 

Sure, we could do that. Bear in mind the ID can be specified before the command sometimes which complicates this; inherited from taskwarrior and I like it. The context also needs to be combined, which might also complicate things.

> While we're at it, i would propose a more step-wise parsing (e.g. we expect an id, look for it, then we expect a filter, parse that, then notes, parse that, etc. rather than running the same loop over and over and setting all kinds of state flags)

Sure, we could do. Unconvinced of its value right now though.


>  The dstask command line is a DSL. It's so simple that it's basically a data structure, but it still needs parsing.

Yes, that's a nice way of thinking about it. After all it does parse into a struct.

Footnote:
I think the `CmdLine` as it exists is really a query that happens to include a command. I will rename it to `Query` in a PR as mentioned in the minutes from the last meeting; in the same PR I want to refactor the command interpretation (and taskSetOpts, which is largely redundant) a bit so we go back to combining the query specified on the command line with the query specified in the context. That should hopefully make it easier to follow and make the thinking behind combining the query/context system as was/is.



-----


I agree with the other concerns.

**Comment by dontlaugh** [comment]
> did you mean main.go instead of commands.go

A bit of both, really. Again, it's so simple that the lines get blurred a bit.

> does a generator have other advantages, such as the ability to generate help text from the language specification

Not on it's own, unfortunately. We'd need to make an api around that, _if_ we adopted such a framework. I wouldn't reach for one, yet. What we have now is fairly simple to unit test.

Perhaps we should focus on a small piece of this problem, maybe by borrowing ideas from urfave/cli (my favorite of these frameworks :) ) [Here's a good example](https://github.com/urfave/cli/blob/master/docs/v2/manual.md#ordering).

If you use this framework, your func main can get quite long, but it is a mostly-declarative and self-documenting list of command and flag declarations, ending in a call to `app.Run(os.Args)`.

We could work towards something like this, and it might not be too difficult. But what would it give us? Well, it might give us a nice way to co-locate our help text and accepted options. And it might help enforce separation of concerns around parsing, taskset construction, taskset manipulation, and output rendering.

Cal has been quite friendly to refactoring the codebase as long as the existing users aren't massively disrupted. [We went back and forth over some concrete designs first](https://github.com/naggie/dstask/issues/41#issuecomment-667707034). He even let me check in several bugs! :sweat_smile:
